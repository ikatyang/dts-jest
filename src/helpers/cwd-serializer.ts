export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
