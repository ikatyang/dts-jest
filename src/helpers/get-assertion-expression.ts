import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
