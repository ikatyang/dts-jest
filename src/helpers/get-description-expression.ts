import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
