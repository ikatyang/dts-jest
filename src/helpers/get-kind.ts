import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
