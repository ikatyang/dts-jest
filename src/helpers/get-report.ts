import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
