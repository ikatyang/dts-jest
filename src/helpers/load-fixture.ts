import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
