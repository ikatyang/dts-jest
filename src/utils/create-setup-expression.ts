import {
  config_namespace,
  package_name,
  runtime_namespace,
  Trigger,
} from '../definitions';

export const create_setup_expression = (triggers: Trigger[]) => {
  const config_expression = `(function () { try { return ${config_namespace} } catch (e) { return {} } })()`;
  return (
    `var ${runtime_namespace} = require(${JSON.stringify(package_name)})` +
    `.setup(module.filename, ${config_expression}, ${JSON.stringify(triggers)})`
  );
};
