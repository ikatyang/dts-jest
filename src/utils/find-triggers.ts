import * as _ts from 'typescript';
import { Trigger, TriggerFooter } from '../definitions';
import { find_trigger_bodies } from './find-trigger-bodies';
import { find_trigger_footers } from './find-trigger-footers';
import { find_trigger_headers } from './find-trigger-headers';
import { get_trigger_body_end_line } from './get-trigger-line';

export const find_triggers = (source_file: _ts.SourceFile, ts: typeof _ts) => {
  const triggers: Trigger[] = [];

  const headers = find_trigger_headers(source_file, ts);
  const bodies = find_trigger_bodies(source_file, headers, ts);
  const footers = find_trigger_footers(source_file, bodies, ts);

  for (let i = 0; i < headers.length; i++) {
    const header = headers[i];
    const body = bodies[i];
    let footer: TriggerFooter | undefined;

    const { line: body_end_line } = source_file.getLineAndCharacterOfPosition(
      body.end,
    );

    if (
      footers.length !== 0 &&
      body_end_line === get_trigger_body_end_line(footers[0].line)
    ) {
      footer = footers.shift();
    }

    triggers.push({
      header,
      body,
      footer,
    });
  }

  return triggers;
};
