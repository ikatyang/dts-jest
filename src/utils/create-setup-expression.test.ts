import * as ts from 'typescript';
import { load_fixture_source_file } from '../helpers/load-fixture';
import { create_setup_expression } from './create-setup-expression';
import { find_triggers } from './find-triggers';

it('should return correctly', () => {
  const source_file = load_fixture_source_file(
    'create-setup-expression/example.ts',
    ts,
  );
  const triggers = find_triggers(source_file, ts);
  expect(create_setup_expression(triggers)).toMatchSnapshot();
});
