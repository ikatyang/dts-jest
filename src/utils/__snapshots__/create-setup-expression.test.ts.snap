// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `"var _dts_jest_runtime_ = require(\\"dts-jest\\").setup(module.filename, (function () { try { return _dts_jest_ } catch (e) { return {} } })(), [{\\"header\\":{\\"line\\":0,\\"flags\\":7,\\"method\\":\\"test\\"},\\"body\\":{\\"start\\":28,\\"end\\":46,\\"one_line_expression\\":\\"Math.max(1, 2, 3)\\",\\"text\\":\\"Math.max(1, 2, 3)\\"},\\"footer\\":{\\"line\\":1,\\"expected\\":\\"3\\"}},{\\"header\\":{\\"line\\":3,\\"flags\\":8,\\"method\\":\\"test\\"},\\"body\\":{\\"start\\":73,\\"end\\":86,\\"one_line_expression\\":\\"Math.abs(-1)\\",\\"text\\":\\"Math.abs(-1)\\"},\\"footer\\":{\\"line\\":4,\\"flag\\":\\"?\\"}}])"`;
